### ISSUE TITLE
Add compatibility to TYPO3 CMS version 8.5 and higher

### ISSUE THREAD

**Issue by angeloprevitali** [jonathanheilmann/ext-jh_magnificpopup#47]
Hello

can you fix the compatibility of the EXT:jh_magnificpopup with the actual TYPO3 CMS version 8.5 and higher?

That would be great.

Thank you very much.

Regards,
Angelo

**Comment by jonathanheilmann** [comment]
I have not done any tests with jh_magnificpopup within TYPO3 CMS version 8.5. Did you do some tests or maybe someone else?
There is a fork by @tollwerk, adding compatibility. Maybe @jkphl could leave some feedback?

**Comment by jkphl** [comment]
@angeloprevitali @jonathanheilmann Well, we did nothing but [remove the version limit to `<7.6.99` from `composer.json`](https://github.com/tollwerk/ext-jh_magnificpopup/commit/57ebacf20d5f3044f2623630e1b453b2c858ef1f) and don't experience any problems so far (with the most recent 8.5.1). However, we don't have any fancy setup, just using plain default, so the devil might well be in the details.

**Comment by jonathanheilmann** [comment]
Thanks for your feedback!

@angeloprevitali Would you like to test the fork? I would appreciate some help and feedback, as I'm highly involved in some other projects and it would take some to do own tests.

**Comment by erredeco** [comment]
I have performed a test with TYPO3 8.7.0 and the most recent jh_magnificpopup (0.6.4)
- click-enlarge on images seem to work (not tested the gallery, yet: just the default functionality)
- when I tried to add a plugin MP for content in the page, the following errors occurred:
1) the icon in wizard was broken
2) after clicking:

> Flex form displayCond on field "settings.mfpOption.focus" on flex form sheet "sMfpConfiguration" > references a field or field / sheet combination "settings.contenttype" that might be defined in given data structure but is not found in data values.
> More information regarding this error might be available online.

**Comment by angeloprevitali** [comment]
@jonathanheilmann We tested the EXT:jh_magnificpopup on a actual 8.7.0 and everything works fine. There are still some little thing that has to be fixed in jh_magnificpopup like the icon as describe above. And also there is a workaround to fix images with FSC. We used it to work with this additional line in TS setup:

`lib.contentElement.settings.media.popup.linkParams.ATagParams.dataWrap = class="magnificpopup" rel="magnificpopup[{field:uid}]" title="{file:current:title}"`

**Comment by erredeco** [comment]
@angeloprevitali  I did not add the line you cited and everything worked out of the box AFAICS.
I just added the constant `styles.content.textmedia.linkWrap.lightboxEnabled = 1` via constant editor

**Comment by erredeco** [comment]
I've found the source of the problem with the plugin: it is inside the flexform file: 
```
<settings.mfpOption.focus>
	<TCEforms>
		<label>LLL:EXT:jh_magnificpopup/Resources/Private/Language/locallang.xlf:pi_flexform.sheetOptions.mfpOption.focus</label>
		<displayCond>FIELD:settings.contenttype:!=:iframe</displayCond>
		<config>
			<type>input</type>
			<size>20</size>
			<max>20</max>
			<eval>trim</eval>
		</config>
	</TCEforms>
</settings.mfpOption.focus>
```
I think that the `<displayCond>FIELD:settings.contenttype:!=:iframe</displayCond>` never worked, but silently failed until TYPO3 7.6.x; from version 8 it  breaks.

**Comment by karinakern** [comment]
I got the extension running on TYPO3 8.7.1 with the help of @erredeco by removing settings.mfpOption.focus in the flexform like mentioned above and also replacing all array_merge() calls in the controller by using a simple addition instead, like mentioned [here](https://docs.typo3.org/typo3cms/extensions/core/Changelog/8.0/Breaking-72342-RemovedDeprecatedCodeFromGeneralUtility.html?highlight=array_merge).

I also had to replace the IMAGE() function with cObjGetSingle('IMAGE', $something) like explained [here.](https://github.com/sir-gawain/tt_news/commit/237f83e6e859e799256df0c76518b077aa5eb1a5)

@jonathanheilmann Do you plan to release a new version that's compatible with TYPO3 8.7.x? Would really appreciate it as I love to work with that extension.

**Comment by jonathanheilmann** [comment]
Thanks for your feedback!
I'm going to do some work this weekend. Maybe there is an official release available in TER on sunday afternoon. Otherwise I'm going to mark and upload a beta version here on github.

**Comment by karinakern** [comment]
Great, thank you! :)

I also added the output of the alt text and title for images linking to an iframe as I needed it for SEO reasons.

`$imageConf['altText'] = $file->getProperty('alternative');
$imageConf['titleText'] = $file->getProperty('title');`
(Added in row 326 in Partials/Magnificpopup/Iframe.html)

Maybe you wanna consider implementing that too...

**Comment by jonathanheilmann** [comment]
There is still some work to do, but I've finished a first working version.
Issues still not resolved:
- Does not work with fluid_styled_content (did it work bevor?)
- RTE links are still failing
- New version has not been tested within TYPO3 CMS 7 LTS

@karinakern I've added your suggestion as it's good for SEO and doesn't break anything available in version 0.6.4

I'd like to ask you to du some tests and leave some feedback.


[jh_magnificpopup_0.7.0-dev.201705141416_201705141416.zip](https://github.com/jonathanheilmann/ext-jh_magnificpopup/files/999329/jh_magnificpopup_0.7.0-dev.201705141416_201705141416.zip)

**Comment by erredeco** [comment]
I am testing v 0.7.0-dev on TYPO3 8.7.1 - the image popup seems to work just adding the constant:
`styles.content.textmedia.linkWrap.lightboxEnabled = 1`
About the other options
-  I have not tested RTE links as CKEditor does not allow to add images
- About jh_magnificpopup for content I have some questions and observations:
1) Display type: inline content
1.1) Q: What's the difference between "Method:ajax" and "Method:inline" ?
1.2) When adding a CE the dropdown "TYPE" remains empty, while "Column" says "[INVALID VALUE ("109")]"
1.3)

2) Display type: Content reference
2.1) Q: What's the difference between "Method:ajax" and "Method:inline" ?

3) I've noticed in all Display types (including "link") that in the source code there is the following:

```
<html xmlns:f="http://xsd.helmut-hummel.de/ns/TYPO3/CMS/Fluid/ViewHelpers" xmlns:jh="http://typo3.org/ns/Heilmann/JhMagnificpopup/ViewHelpers"></html>

```

**Comment by jonathanheilmann** [comment]
Thanks for your feedback. I've updated 0.7.0-dev fixing some things.

1.1 & 2.1:
The difference is the way the content in MagnificPopup is loaded: Method "inline" adds the content as hidden content to your main-page. Thus it's loaded instantly. But if there is much content and/or it's not often opened by visitors, it's an overload.
In this case it's recommended to use Method "ajax" sending a new request to your server and just loading the required content (using a dedicated typeNum or eID request). Thus serverload is optimized but another request is done, forcing the visitor to wait for a result.

1.2:
When using the attached version, is TYPE still empty? "Column" is nasty: Inline Content should not be displayed in default columns, as is should no be visible by default. So it's moved to it's own column. Column type 109 is added by the extension, but it's not visible, as the itemProcFunc of field colPos removes it, as it's not configured in BackendLayout. A solution could be, to add colPos 109 to your BackendLayout or adding an own itemProcFunc. But adding an own itemProcFunc may conflict with other extensions like gridelements. 

[jh_magnificpopup_0.7.0-dev.201705150723_201705150723.zip](https://github.com/jonathanheilmann/ext-jh_magnificpopup/files/1000148/jh_magnificpopup_0.7.0-dev.201705150723_201705150723.zip)

**Comment by karinakern** [comment]
I tested the latest version you uploaded on my TYPO3 8.7.1 setting and so far it works fine for me.

What exactly do you mean by "Does not work with fluid_styled_content"? I'm using fluid_styled_content and did not experience any errors so far.

As far as I remember RTE links used to work by selecting the special class for magnific popup, right? Is that what's still missing/not working with fluid? Cause by default it seems there are no classes predefined for links at all any more in TYPO3 8 / CKEditor.

Thanks a lot for your effort!!

**Comment by jonathanheilmann** [comment]
Actually I'm testing the new version within three setups:
1. TYPO3 8.7.1 with introduction package and css_styled_content
2. TYPO3 8.7.1 with introduction package and fluid_styled_content
3. TYPO3 8.7.1 with raw fluid_styled_content just creating a simple page without any styles

The first one works great, the other two are failing when rendering inline or ajax content.
`inline` renders the content in controller, but the output of the complete contentElement is skipped and looks like this:
```
<div id="c198" class="frame frame-default frame-type-list frame-layout-0">
  <header>
    <h2 class="">
      inline
    </h2>
  </header>
  <div id="c" class="frame frame- frame-type- frame-layout-"></div>
</div>
```
`ajax` renders all contentElements of the page instead of just the elements related to the plugin contentElement.

I'll try to provide a public test environment on Thursday.

**Comment by Ulfried** [comment]
Hi Jonathan,

fluid_styled_content comes with the CE 'Image' again, and it uses a different wrapper class than the CEs textpic / textmedia. Could you add `.ce-image` to the default selectors in /Configuration/TypoScript/Default/setup.txt (in line 54) please:
`                1.value = .csc-textpic-imagewrap,.ce-textpic,.ce-image`

Best,
Ulfried

**Comment by Anubiso** [comment]
Any updates?

**Comment by christophlehmann** [comment]
We use it together with CKEditor. Since 8.7.5 it breaks the CKEditor: Text styles get lost on saving. I identified the lines https://github.com/jonathanheilmann/ext-jh_magnificpopup/blob/master/ext_localconf.php#L43 to https://github.com/jonathanheilmann/ext-jh_magnificpopup/blob/master/ext_localconf.php#L47 are breaking it. (Later in that file again in "alpha features")

I suggest not to load the RTE configuration by default and better supply it like done in news extension

```
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::registerPageTSConfigFile(
    'news',
    'Configuration/TSconfig/Page/news_only.txt',
    'EXT:news :: Restrict pages to news records');
```
or load it only when `\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::isLoaded('ckeditor')` is `false`

**Comment by jonathanheilmann** [comment]
It's been a while since I've updated this issue; now, there are some good news and a bad one:

I've managed to solve the missing content bug for EXT:bootstrap_package, I'm going to clean up the code, merge some pull requests and solve as much issues as possible within the next days.

The bad news:
The bugfix requires a breaking change within the templates and I'm going to break some other things. For those who overrides the template: I'm sorry, but you will need to update yours.

Cheers and stay tuned!
